File: cpp/leetcode-stairs.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int climbStairs(int n) {
        if (n <= 3) return n;
        vector<int> dp(n + 1, 0);
        return fib(n, dp);
    }

    int fib(int n, vector<int>& dp) {
        if (n <= 3) return n;
        if (dp[n]) return dp[n];
        dp[n] = fib(n - 1, dp) + fib(n - 2, dp);
        return dp[n];
    }
};

File: rs/leetcode-stairs.rs
impl Solution {
    pub fn climb_stairs(n: i32) -> i32 {
        let mut dp:Vec<i32> = vec![0; (n + 1) as usize];
        Solution::fib(n, &mut dp)
    }

    fn fib(n:i32, dp:&mut Vec<i32>) -> i32{
        if n <=3 {return n;}
        if dp[n as usize] != 0 {return dp[n as usize];}
        dp[n as usize] = Solution::fib(n - 1, dp) + Solution::fib(n - 2, dp);
        dp[n as usize]
    }
}